gobblin-modules/google-ingestion/src/main/java/gobblin/ingestion/google/webmaster/GoogleWebmasterDataFetcherImpl.java

```diff
@@ -0,0 +1,383 @@
+      if (actualSize != expectedSize) {
+        LOG.warn(
+            String.format("Expected page size for country-%s is %d, but only able to get %d", country, expectedSize,
+                actualSize));
+      }
+      LOG.info(String.format("A total of %d pages fetched for property %s at country-%s from %s to %s", actualSize,
+          _siteProperty, country, startDate, endDate));
+    }
+
+    ArrayDeque<ProducerJob> jobs = new ArrayDeque<>(actualSize);
+    for (String page : allPages) {
+      jobs.add(new SimpleProducerJob(page, startDate, endDate));
+    }
+    return jobs;
+  }
+
+  private int getPagesSize(final String startDate, final String endDate, final String country,
+      final List<Dimension> requestedDimensions, final List<ApiDimensionFilter> apiDimensionFilters)
+      throws IOException {
+    int REQUESTS_COUNT_EACH_ROUND = 4;
+    //Each round will give you 20K(4*5000) pages. Doing 100 rounds can give 2 million pages.
+    //Normally 1 week has less than 200K pages. So 2 million pages are more than enough.
+    int MAXIMUM_ROUNDS = 100;
+    final ExecutorService es = Executors.newFixedThreadPool(REQUESTS_COUNT_EACH_ROUND,
+        ExecutorsUtils.newDaemonThreadFactory(Optional.of(LOG), Optional.of(this.getClass().getSimpleName())));
+
+    int startRow = 0;
+    int r = 0;
+    while (r < MAXIMUM_ROUNDS) {
+      r++;
+      List<Future<Integer>> results = new ArrayList<>(REQUESTS_COUNT_EACH_ROUND);
+
+      for (int i = 0; i < REQUESTS_COUNT_EACH_ROUND; ++i) {
+        startRow += GoogleWebmasterClient.API_ROW_LIMIT;
+        final int start = startRow;
+        //Submit the job.
+        Future<Integer> submit = es.submit(new Callable<Integer>() {
+          @Override
+          public Integer call() {
+            LOG.info(String.format("Getting page size from %s...", start));
+            while (true) {
+              if (Thread.interrupted()) {
+                LOG.error(String.format(
+                    "Interrupted while trying to get the size of all pages for %s. Current start row is %d.", country,
+                    start));
+                return -1;
+              }
+              try {
+                List<String> pages =
+                    _client.getPages(_siteProperty, startDate, endDate, country, GoogleWebmasterClient.API_ROW_LIMIT,
+                        requestedDimensions, apiDimensionFilters, start);
+                if (pages.size() < GoogleWebmasterClient.API_ROW_LIMIT) {
+                  return pages.size() + start;  //Figured out the size
+                } else {
+                  return -1;
+                }
+              } catch (IOException e) {
+                LOG.info(String.format("Getting page size from %s failed. Retrying...", start));
+              }
+              try {
+                Thread.sleep(200);
+              } catch (InterruptedException e) {
+                LOG.error(e.getMessage());
+                LOG.error(String.format(
```

Comment:
Please merge the two LOG.error in one, logs are mostly interleaved with each other due to multiple threads logging simultaneously